### PR TITLE
Allows the data fetcher of last resort to be specified

### PR DESCRIPTION
**Pull request by bbakerman**
#536 - the data fetcher used when no other data fetcher has been provided

### CODE REVIEW

**Comment by A1m7331** [comment]
I would call it DefaultDataFetcher, last resort sound kind of strange :)

**Comment by tklovett** [comment]
If I'm reading this correctly, this only allows one to specify a "global" default data fetcher for _all_ fields. It would also be nice to be able to specify a default data fetcher at the type level.

For example, we want to use an `io.vavr.Either<Error, Success>` as the Java backing of a `union ErrorOrSuccess = Error | Success` for handling common user errors (e.g. "That username is already taken!"

Since the `DataFetcher<Either<Error, Success>>` doesn't return an Object with properties, the default `PropertyDataFetcher` does not work for pulling scalar fields out of the object. We've created an "`EitherPropertyDataFetcher`" which 1) maps the `Either` to the `Left` or the `Right` projection and b) delegates to a `PropertyDataFetcher` to fetch the property from the `Left` or the `Right` projection.

But this means that we have we have to register the `EitherPropertyDataFetcher` as the `.dataFetcher(...)` for _every_ field on the type:

```
    private TypeRuntimeWiring.Builder ErrorOrSuccess() {
        return TypeRuntimeWiring.newTypeWiring("ErrorOrSuccess")
                .typeResolver(new EitherTypeResolver<Error, Success>("Error", "Success"));
    }

    private TypeRuntimeWiring.Builder Error() {
        return TypeRuntimeWiring.newTypeWiring("Error")
                .dataFetcher("message", EitherPropertyDataFetcher.forProperty("message"))
                .dataFetcher("successful", EitherPropertyDataFetcher.forProperty("successful"));
    }

    private TypeRuntimeWiring.Builder Success() {
        return TypeRuntimeWiring.newTypeWiring("Success")
                .dataFetcher("id", EitherPropertyDataFetcher.forProperty("id"))
                .dataFetcher("successful", EitherPropertyDataFetcher.forProperty("successful"));
    }
```

Ideally, we would be able to register one `DataFetcher` to be used for every property. e.g.:

```
    private TypeRuntimeWiring.Builder Error() {
        return TypeRuntimeWiring.newTypeWiring("Error")
                .fieldDataFetcher(new EitherPropertyDataFetcher());
    }
```

**Comment by bbakerman** [comment]
@tklovett - we have the WiringFactory if you want to provide the same fetcher always to a specific type.  Its less visible (in a code sense) than the "manual wiring" but works the same

Just respond to `providesDataFetcher` to all fields of a certain type

File: src/main/java/graphql/schema/idl/SchemaGenerator.java [hunk 1]
```diff
@@ -476,12 +476,13 @@ private DataFetcher buildDataFetcher(BuildContext buildCtx, TypeDefinition paren
...
+            dataFetcher = runtimeWiring.getDataFetcherForType(parentTypeName).get(fieldName);
             if (dataFetcher == null) {
-                //
-                // in the future we could support FieldDateFetcher but we would need a way to indicate that in the schema spec
-                // perhaps by a directive
-                dataFetcher = new PropertyDataFetcher(fieldName);
+                dataFetcher = runtimeWiring.getDefaultDataFetcherForType(parentTypeName);
+                if (dataFetcher == null) {
+                    dataFetcher = wiringFactory.getDefaultDataFetcher(wiringEnvironment);
+                    assertNotNull(dataFetcher, "The WiringFactory indicated MUST provide a data fetcher of last resort as part of its contract");
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: naming: here it is still called `last resort`

File: src/main/java/graphql/schema/idl/TypeRuntimeWiring.java [hunk 2]
```diff
@@ -115,6 +122,21 @@ public Builder dataFetchers(Map<String, DataFetcher> dataFetchersMap) {
         }
 
         /**
+         * All fields in a type need a data fetcher of some sort and this method is called to prove the data fetcher
+         * of last resort, that is the one that will be used for this type if no specific one has been provided
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/TypeRuntimeWiring.java]: naming: here it is still called `last resort`

File: src/main/java/graphql/schema/idl/WiringFactory.java [hunk 3]
```diff
@@ -79,4 +80,15 @@ default boolean providesDataFetcher(FieldWiringEnvironment environment) {
     default DataFetcher getDataFetcher(FieldWiringEnvironment environment) {
         return assertShouldNeverHappen();
     }
+
+    /**
+     * All fields need a data fetcher of some sort and this  method is called to prove the data fetcher
+     * of last resort, that is the one that will be used if no specific one has been provided
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/WiringFactory.java]: naming: here it is still called `last resort`

File: src/test/groovy/graphql/schema/idl/WiringFactoryTest.groovy [hunk 4]
```diff
@@ -194,4 +194,50 @@ class WiringFactoryTest extends Specification {
 
         true // assertions in callback
     }
+
+    def "last resort data fetcher is used"() {
```
Comment (andimarek) [inline src/test/groovy/graphql/schema/idl/WiringFactoryTest.groovy]: naming: here it is still called `last resort`

**Comment by bbakerman** [comment]
> I am not sure about the WiringFactory method:

The reason I put this in is that previously if not one answered YES to providing a data fetcher then PropertyDataFetcher was used.

I tried to replace that by saying you MUST wire in a default type of data fetcher and then providing a default implementation that used PropertyDataFetcher anyway.

So we need a place to provide the "global" proeprty fetcher.  This could be on RuntimeWiring (not per type though) or as I chose in the WiringFactory

See https://github.com/graphql-java/graphql-java/issues/536 for more background from other people

I am happy to make the "global" data fetcher on RuntimeWiring rather than WiringFactory.  In either case it will default to PropertyDataFetcher

**Comment by bbakerman** [comment]
Actually it cant be global on RuntimeWiring since the getDefaultDF requires the "field" to be fetched to be named as input

Hence it needs to the `FieldWiringEnvironment` input to make a proper decision.

And thats what WiringFactory does

**Pull request merged**